### PR TITLE
loadbalancingexporter: add support for logs - 1/5

### PR DESCRIPTION
**Pull request by BinaryHexer**
**Description:** <Describe what has changed.>

In preparation for introducing log exporter to the loadbalancingexporter.

- In `pkg`
  - Add `batchpersignal` package.

- In `pkg/batchpersignal`
  - Move `Split` function from `batchpertrace` and rename as `SplitTraces`.
  - Add `SplitLogs` function

- In `pkg/batchpertrace`
  - Deprecate the package.
  - Make `Split` function depend on `batchpersignal.SplitTraces`.

- In `exporter/loadbalancingexporter`
  - Replace references to `batchpertrace.Split` with `batchpersignal.SplitTraces`.
  - Rename `newExporter` to `newTracesExporter`
  - Rename `exporterImp` to `traceExporterImp`
  - Rename `exporter.go` and `exporter_test.go` to `traces_exporter.go` and `traces_exporter_test.go` respectively.

**Link to tracking Issue:** https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/2470

Testing: Unit tests

Documentation: none (as part of this PR)

Signed-off-by: Vijayender Joshi vijayeder.joshi@gmail.com

### CODE REVIEW

File: pkg/batchpertrace/batchpertrace.go [hunk 1]
```diff
@@ -63,3 +63,51 @@ func Split(batch pdata.Traces) []pdata.Traces {
...
+	// for each log in the resource logs, we group them into batches of rl/ill/traceID.
+	// if the same traceID exists in different ill, they land in different batches.
+	var result []pdata.Logs
+
+	for i := 0; i < batch.ResourceLogs().Len(); i++ {
+		rs := batch.ResourceLogs().At(i)
+
+		for j := 0; j < rs.InstrumentationLibraryLogs().Len(); j++ {
+			// the batches for this ILL
+			batches := map[[16]byte]pdata.ResourceLogs{}
```
Comment (jpkrohling) [inline pkg/batchpertrace/batchpertrace.go:78]: I think a pdata.TraceID can be used here now. It wasn't possible in the first implementation, IIRC.

File: pkg/batchpertrace/batchpertrace.go [hunk 2]
```diff
@@ -63,3 +63,51 @@ func Split(batch pdata.Traces) []pdata.Traces {
...
+
+			ill := rs.InstrumentationLibraryLogs().At(j)
+			for k := 0; k < ill.Logs().Len(); k++ {
+				log := ill.Logs().At(k)
+				key := log.TraceID().Bytes()
```
Comment (jpkrohling) [inline pkg/batchpertrace/batchpertrace.go:83]: I would assume that not every log has a TraceID. Will it then be null?

File: pkg/batchpertrace/batchpertrace_test.go [hunk 3]
```diff
@@ -56,3 +56,39 @@ func TestSplitDifferentTracesIntoDifferentBatches(t *testing.T) {
...
+	rl.InstrumentationLibraryLogs().Resize(1)
+
+	// the first ILS has two spans
+	ill := rl.InstrumentationLibraryLogs().At(0)
+	library := ill.InstrumentationLibrary()
+	library.SetName("first-library")
+	ill.Logs().Resize(2)
+	firstLog := ill.Logs().At(0)
+	firstLog.SetName("first-batch-first-log")
+	firstLog.SetTraceID(pdata.NewTraceID([16]byte{1, 2, 3, 4}))
```
Comment (jpkrohling) [inline pkg/batchpertrace/batchpertrace_test.go:74]: Could you add a test for a log entry that doesn't have a trace ID?

**Review by jpkrohling — COMMENTED** [review]
Looks interesting, but have you got clearance from maintainers? I believe they are accepting only changes that are relevant for the GA at the moment.

File: pkg/batchpertrace/batchpertrace.go (hunk 1 above)
Comment (bogdandrutu) [reply pkg/batchpertrace/batchpertrace.go:78]: Yes because we were exposing as byte slice not array

**Comment by BinaryHexer** [comment]
@jpkrohling Thanks for your helpful comments!

Sorry if I missed anything (I did my best to check the CONTRIBUTING.MD) but is there any specific process I need to follow to get clearance from the maintainers for this?

I was hoping since this is not an entirely new component, just adding support for another signal it would be okay.

File: pkg/batchpertrace/batchpertrace.go [hunk 4]
```diff
@@ -16,8 +16,8 @@ package batchpertrace
 
 import "go.opentelemetry.io/collector/consumer/pdata"
 
-// Split returns one pdata.Traces for each trace in the given pdata.Traces input. Each of the resulting pdata.Traces contains exactly one trace.
-func Split(batch pdata.Traces) []pdata.Traces {
+// SplitTraces returns one pdata.Traces for each trace in the given pdata.Traces input. Each of the resulting pdata.Traces contains exactly one trace.
```
[the remark this answers is not in the document]
Comment (BinaryHexer) [reply pkg/batchpertrace/batchpertrace.go]: Done in [`b8da467` (#2549)](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/2549/commits/b8da467d196f280f8b52db433e350e08026cb8a5)

File: pkg/batchpertrace/batchpertrace.go (hunk 1 above)
Comment (BinaryHexer) [reply pkg/batchpertrace/batchpertrace.go]: Fixed it both for `SplitLogs` and `SplitTraces` function.

File: pkg/batchpertrace/batchpertrace.go (hunk 2 above)
Comment (BinaryHexer) [reply pkg/batchpertrace/batchpertrace.go]: `.IsEmpty()` returns true for the TraceID, but TraceID itself is not null.

File: pkg/batchpertrace/batchpertrace_test.go (hunk 3 above)
Comment (BinaryHexer) [reply pkg/batchpertrace/batchpertrace_test.go]: Done here: https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/2549/files#diff-5b927bd0ee2bf7e5ee21db9bf4a4a6e4406b4f75e83ff616c5d176117ef852d5R80

**Comment by tigrannajaryan** [comment]
@BinaryHexer this is the [part](https://github.com/open-telemetry/opentelemetry-collector/blob/main/CONTRIBUTING.md#before-you-start) from core contributing guide that @jpkrohling refers to:
>If you would like to work on something that is not listed as an issue (e.g. a new feature or enhancement) please first read our vision and roadmap to make sure your proposal aligns with the goals of the Collector, then create an issue and describe your proposal. It is best to do this in advance so that maintainers can decide if the proposal is a good fit for this repository. This will help avoid situations when you spend significant time on something that maintainers may decide this repo is not the right place for.

You submitted an issue which is great, but the maintainers haven't yet seen and given a green flag.

In this case I think we are fine. I believe the change is simple and since you already started if @jpkrohling is willing to review as a codeowner then we are good to go.

We are much more reluctant about non-trivial new feature proposal at the moment, especially in the core repo and the ones that require a design and significant time investment from maintainers. This is neither, so your contribution is very welcome, provided that @jpkrohling has the cycles to review the PRs.

**Comment by BinaryHexer** [comment]
@tigrannajaryan Thank you for explaining it to me in detail. In the future I will keep this in mind for any future issues / features. 🙇

**Review by jpkrohling — APPROVED** [review]
Looks great!

**Comment by bogdandrutu** [comment]
Please rebase

**Comment by jpkrohling** [comment]
Rebased, please review/merge

**Comment by bogdandrutu** [comment]
@jpkrohling I trust you 🗡️

**Pull request merged**